### PR TITLE
Configurable namespaces for CRs

### PR DESCRIPTION
**Pull request by shlomitk1**
Fixes https://github.com/fybrik/fybrik/issues/1987

### CODE REVIEW

File: pkg/environment/environment.go [hunk 1]
```diff
@@ -95,6 +97,22 @@ func GetApplicationNamespace() string {
 	return os.Getenv(ApplicationNamespace)
 }
 
+func GetSystemNamespace() string {
```
Comment (roytman) [inline pkg/environment/environment.go:100]: GetIntCRDsNamespace ?

File: pkg/environment/environment.go [hunk 2]
```diff
@@ -95,6 +97,22 @@ func GetApplicationNamespace() string {
...
+	systemNamespace := os.Getenv(SystemNamespace)
+	if systemNamespace == "" {
+		systemNamespace = GetControllerNamespace()
+	}
+	return systemNamespace
+}
+
+func GetAdminNamespace() string {
```
Comment (roytman) [inline pkg/environment/environment.go:108]: GetAdminCRDsNamespace ?

File: manager/testdata/notebook/write-flow/setup.sh [hunk 3]
```diff
@@ -7,20 +7,21 @@ kubectl create namespace fybrik-notebook-sample
 kubectl config set-context --current --namespace=fybrik-notebook-sample
 
 FYBRIK_NAMESPACE=fybrik-system
+ADMIN_NAMESPACE?=fybrik-system
```
Comment (roytman) [inline manager/testdata/notebook/write-flow/setup.sh:10]: ADMIN_CRDS_NAMESPACE ?

File: manager/main.go [hunk 4]
```diff
@@ -71,13 +71,15 @@ func run(namespace, metricsAddr, healthProbeAddr string, enableLeaderElection bo
 	setupLog.Info().Msg("Application namespace: " + applicationNamespace)
 
 	systemNamespaceSelector := fields.SelectorFromSet(fields.Set{"metadata.namespace": environment.GetSystemNamespace()})
+	adminNamespaceSelector := fields.SelectorFromSet(fields.Set{"metadata.namespace": environment.GetAdminNamespace()})
```
Comment (roytman) [inline manager/main.go:74]: adminCRDsNamespaceSelector := fields.SelectorFromSet(fields.Set{"metadata.namespace": environment.GetAdminCRDsNamespace()})

File: pkg/environment/environment.go [hunk 5]
```diff
@@ -35,6 +35,8 @@ const (
 	ModuleNamespace                   string = "MODULES_NAMESPACE"
 	ControllerNamespace               string = "CONTROLLER_NAMESPACE"
 	ApplicationNamespace              string = "APPLICATION_NAMESPACE"
+	AdminNamespace                    string = "ADMIN_NAMESPACE"
```
Comment (roytman) [inline pkg/environment/environment.go:38]: AdminCRDsNamespace ..

**Review by roytman — COMMENTED** [review]
just an issue with names, see comments. I think the current names can mislead.

File: charts/fybrik/values.yaml [hunk 6]
```diff
@@ -25,6 +25,16 @@ modulesNamespace:
 # The namespace must already exist
 applicationNamespace: ""
 
+# The namespace where Fybrik should allocate and monitor its internal CRs (Blueprints and Plotters)
+# If not defined, the control plane namespace will be used.
+# Otherwise, the namespace must already exist.
+systemNamespace: ""
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/values.yaml]: systemNamespace -> internalCRsNamespace
adminNamespace -> adminCRsNamespace
fybrik-internal-crs-role/rb
fybrik-admin-crs-role/rb
Also, renamed fybrik-admin-cr to fybrik-applications-cr, similar to fybrik-applications-role, as both refer to fybrikapplications.

File: charts/fybrik/templates/fybrik-cr-role.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
+{{- if .Values.manager.enabled }}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/templates/fybrik-cr-role.yaml]: fybrik-admin-crs-role/rb

File: charts/fybrik/values.yaml [hunk 8]
```diff
@@ -25,6 +25,16 @@ modulesNamespace:
...
+
+# The namespace where Fybrik should watch admin CRs (FybrikModules and FybrikStorageAccounts)
+# If not defined, the control plane namespace will be used.
+# Otherwise, the namespace must already exist.
+adminNamespace: ""
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/values.yaml]: adminCRsNamespace
fybrik-admin-crs-role/rb

File: charts/fybrik/templates/fybrik-system-role.yaml [hunk 9]
```diff
@@ -3,11 +3,13 @@ apiVersion: rbac.authorization.k8s.io/v1
 kind: Role
 metadata:
   name: {{ template "fybrik.fullname" . }}-system-role
+{{- if .Values.systemNamespace }}
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply charts/fybrik/templates/fybrik-system-role.yaml]: fybrik-internal-crs-role/rb

File: manager/controllers/app/blueprint_controller.go [hunk 10]
```diff
@@ -425,7 +425,7 @@ func (r *BlueprintReconciler) getExpectedResults(kind string) (*fapp.ResourceSta
 	ctx := context.Background()
 
 	var moduleList fapp.FybrikModuleList
-	if err := r.List(ctx, &moduleList, client.InNamespace(environment.GetSystemNamespace())); err != nil {
+	if err := r.List(ctx, &moduleList, client.InNamespace(environment.GetAdminNamespace())); err != nil {
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/blueprint_controller.go]: GetAdminCRsNamespace

File: site/docs/contribute/modules.md [hunk 11]
```diff
@@ -159,7 +159,7 @@ metadata:
   labels:
     name: "<module name>" # the name of your new module
     version: "<semantic version>"
-  namespace: fybrik-system  # control plane namespace. Always fybrik-system
+  namespace: <module namespace>  # module namespace, configurable in values.yaml. The default is "fybrik-system".
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply site/docs/contribute/modules.md]: done

File: pkg/environment/environment.go (hunk 1 above)
Comment (shlomitk1) [reply pkg/environment/environment.go]: GetInternalCRsNamespace

File: pkg/environment/environment.go (hunk 2 above)
Comment (shlomitk1) [reply pkg/environment/environment.go]: GetAdminCRsNamespace

File: manager/testdata/notebook/write-flow/setup.sh (hunk 3 above)
Comment (shlomitk1) [reply manager/testdata/notebook/write-flow/setup.sh]: ADMIN_CRS_NAMESPACE

File: pkg/environment/environment.go (hunk 5 above)
Comment (shlomitk1) [reply pkg/environment/environment.go]: AdminCRsNamespace

File: manager/main.go (hunk 4 above)
Comment (shlomitk1) [reply manager/main.go]: changed

**Review by roytman — APPROVED** [review]

**Pull request merged**